Use reckless calls where possible, for performance

bemasc opened the pull request:
Use reckless calls where possible for performance.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-lib/182)

<!-- Reviewable:end -->

bemasc committed:
NOT FOR COMMIT UNTIL NEXT FREEDOM CUT

Use reckless calls where possible for performance.

trevj commented on src/churn-pipe/churn-pipe.ts:
```diff
@@ -194,14 +194,12 @@ class Pipe {
-      to.port).then(() => {
-        return Promise.resolve();
-      });
+      to.port)
```
missing semi-colon

bemasc replied:
Done.

trevj replied:
:+1:

trevj commented on src/net/tcp.ts:
```diff
@@ -370,8 +369,8 @@ export class Connection {
-        return this.counter_.wrap(this.connectionSocket_.write.bind(
-                this.connectionSocket_, buffer));
+        this.connectionSocket_.write.reckless(buffer);
+        return Promise.resolve<void>();
```
can't we use `setSyncHandler`, then we don't have to return a dummy promise?

bemasc replied:
Good point! Done.

trevj replied:
:+1:

trevj commented on third_party/freedom-typings/freedom-common.d.ts:
```diff
@@ -134,6 +134,22 @@ declare module freedom {
+  // This generic interface represents any freedom method. Its purpose is to extend
+  // the basic definition to include the reckless call method, which does not
+  // produce a reply message.
+  interface Method1<T,R> {
```
the number suffix reminds me of a certain Java API i've worked with in the past...

seems okay, but how about `Reckless1`, `Reckless2`, ...?

bemasc replied:
Yeah, `Method` is not a great name ... but I'm not sure `Reckless` is right either, because _every_ method supports this pattern.

trevj replied:
Right, since it applies to every function I guess this makes sense.

trevj replied:
:+1:

trevj commented on third_party/freedom-typings/rtcdatachannel.d.ts:
```diff
@@ -34,8 +34,8 @@ declare module freedom_RTCDataChannel {
     setBinaryType(type:string) : Promise<void>;
-    send(message:string) : Promise<void>;
-    sendBuffer(message:ArrayBuffer) : Promise<void>;
+    send :freedom.Method1<string,void>;
```
so, do all void-returning methods have a reckless counterpart?

bemasc replied:
No, all methods have a void-returning reckless counterpart :-).

trevj replied:
Cool, got it.

trevj replied:
:+1:

trevj commented:
Great. Thanks for pushing this through on the freedomjs side.

Do you have a (very is fine) rough idea of the throughput boost this provides?

bemasc committed:
Cleanup in response to comments.

bemasc commented:
Just tested, seeing 10% boost on CHURN, or maybe a hair more.

trevj commented:
:+1:

bemasc committed:
Update deps to freedom version that support .reckless.

bemasc committed:
Add reckless pause and resume.

Recent testing has reminded us that these are on the critical path,
and likely to become more critical as we improve the efficiency of
other aspects of the network stack.

This change also fixed some compilation issues.

bemasc committed:
More compilation fixes.

bemasc commented:
I extended this change to include pause() and resume().  Now seeing about a 15% boost.

trevj commented on src/net/udpclient.ts:
```diff
@@ -36,7 +37,20 @@ class UdpClient {
     // TODO: not sure what else this should be?
     return this.socket.bind('127.0.0.1', 0)
-        .then((endpoint:net.Endpoint) => {
+        .then((resultCode:number) => {
```
This file has been deleted, so you might want to re-base :-)

Sorry I didn't do this earlier...

bemasc replied:
Thanks! Done.

trevj commented on src/net/tcp.ts:
```diff
@@ -369,9 +368,8 @@ export class Connection {
-        return this.counter_.wrap(this.connectionSocket_.write.bind(
-                this.connectionSocket_, buffer));
+      this.dataToSocketQueue.setSyncHandler((buffer:ArrayBuffer) : void => {
+        this.connectionSocket_.write.reckless(buffer);
```
seems like a candidate for `bind` but obviously this is fine

bemasc replied:
Done.  (Doesn't even need bind, actually...)

trevj commented on src/net/tcp.ts:
```diff
@@ -475,7 +473,7 @@ export class Connection {
    * Sends a message that is pre-formatted as an arrayBuffer.
    */
-  public send = (msg :ArrayBuffer) : Promise<freedom_TcpSocket.WriteInfo> => {
+  public send = (msg :ArrayBuffer) : Promise<void> => {
```
is there any need to return anything here now?

bemasc replied:
I added a comment here noting that the Promise can reject under certain circumstances.  I think that might make the return worthwhile.

trevj commented on src/rtc-to-net/rtc-to-net.ts:
```diff
@@ -474,24 +474,21 @@ import logging = require('../logging/logging');
       this.channelReceivedBytes_ += data.buffer.byteLength;
 
-      return this.tcpConnection_.send(data.buffer);
+      this.tcpConnection_.send(data.buffer);
```
question: what happens when a reckless call raises an exception?

bemasc replied:
Reckless calls never raise an exception (that's why they're reckless!).

If the method call throws or rejects in the core environment, it should trigger a freedom log message, so we should still see it in the logs.

trevj replied:
OK. That seems like something we might want to change about `reckless`.

trevj commented on src/webrtc/datachannel.ts:
```diff
@@ -285,28 +285,25 @@ export class DataChannelClass implements DataChannel {
+        throw new Error('Bad data: ' + JSON.stringify(data));
       }
     // Can raise NetworkError if channel died, for example.
     } catch (e) {
```
i just noticed this nasty "log and re-throw" anti-pattern -- might be an opportunity now to remove it?

bemasc replied:
Done.

trevj commented:
Looking very very good. Just a couple of things...mostly, I'm wondering how exceptions behave with reckless calls.

bemasc committed:
Cleanup in response to comments.

trevj commented:
:+1:
